financial#173: Show 'Pay Now' on user dashboard for partially paid contributions

MegaphoneJon opened the pull request:
https://lab.civicrm.org/dev/financial/-/issues/173

Overview
----------------------------------------
The user dashboard has a "Pay Now" button next to contributions of type "Pending" but not those of type "Partially Paid".

Before
----------------------------------------
![Selection_1102](https://user-images.githubusercontent.com/1796012/119159666-78e8a180-ba25-11eb-8f80-b4dfd512047e.png)

After
----------------------------------------
![Selection_1101](https://user-images.githubusercontent.com/1796012/119159683-7d14bf00-ba25-11eb-8cd7-03e6893c4fc4.png)

Comments
----------------------------------------
@magnolia61 Am I forgetting something important about this?

MegaphoneJon commented:
I recognize that despite being form-level, it is and should be testable.  If I get "Concept Approved" I'll add the test.

eileenmcnaughton commented:
Yep - I agree with this change & @magnolia61 is the right person to test!

magnolia61 commented:
Hello @MegaphoneJon here is a slightly updated PR that was based on previous work by https://github.com/civicrm/civicrm-core/pull/20382 ny @jitendrapurohit 

This PR also outputs correct table layout display and a balance field.

I will also try to test the functionality itself the coming days

eileenmcnaughton commented:
@MegaphoneJon from ^^ - this would definitely be mergeable with a test - potentially you could pull in some of @magnolia61's changes as he is losing the jenkins battle at the moment :-)

MegaphoneJon commented:
Thanks to both of you for the attention on this.  @magnolia61 - unless you tell me otherwise, I'm going to merge your changes into this PR and get to the bottom of the test failure in the next day or two.

eileenmcnaughton commented:
hint @magnolia61 you don't want to tell him not to :-)

magnolia61 commented:
@eileenmcnaughton I created a second PR https://github.com/civicrm/civicrm-core/pull/20388#issuecomment-846697394 but that got stuck on the test. Would be glad if @MegaphoneJon integrates it. Thanks for working on this.

MegaphoneJon commented:
test this

MegaphoneJon commented:
test this please

MegaphoneJon commented:
@magnolia61 This PR is now #20388 exactly, plus some test fixes.

If you look at the tests, they should speak for themselves - but they ensure that the correct headers appear in the correct order, and all the cells in the row are present.  Since we added headers and cells, the tests were failing, so I corrected the tests to reflect the new headers/cell.

eileenmcnaughton commented:
@MegaphoneJon jenkins is given an error - maybe do a pull --rebase & force push? We seem to have been seeing this a lot lately - not sure why

magnolia61 commented:
![image](https://user-images.githubusercontent.com/2195908/119752890-11f04000-be9e-11eb-8d3b-f76f53f3eb79.png) yeah :-)

eileenmcnaughton commented:
It's a green light from jenkins - were you going to add in a test @MegaphoneJon ?

MegaphoneJon commented:
Oh, right!  Tomorrow.  I should be able to extend the current tests to cover this.

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
Fail looks unrelated - @magnolia61 did you test the final version?

magnolia61 commented:
I can confirm that the pay now button now appears when it should (at contributions with status 'pending' and with 'partially paid'.

Also the balance field works nicely showing details about amount still due, and amount pending refund.

![image](https://user-images.githubusercontent.com/2195908/120084078-bd330c00-c0cd-11eb-81d3-07b2a3709900.png)

magnolia61 commented:
Houston... When I tested I looked at the appearance of the buttons, and also if the contribution would be ok (the remaining amount). Unfortunately I noticed the payment amount is not recorded well. During the payment itself the proper remaining amount is charged, but the payment is recorded in civicrm as being the full amount of the contribution, disrupting the total balance. I opend up an issue for this. Possibly a very easy needle somewhere, but in which haystack? https://lab.civicrm.org/dev/financial/-/issues/174

eileenmcnaughton commented:
My take on this - also on the issue - is that the url should point to the add payment form civicrm/contact/view/contribution?reset=1&action=add&cid=202&context=contribution&mode=live not the contribution page
